src/main/java/dev/codesupport/web/common/service/http/SpringHttpClient.java

```diff
@@ -0,0 +1,93 @@
+package dev.codesupport.web.common.service.http;
+
+import dev.codesupport.web.common.exception.HttpRequestException;
+import dev.codesupport.web.common.exception.InternalServiceException;
+import dev.codesupport.web.common.util.MappingUtils;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Component
+@EqualsAndHashCode
+public class SpringHttpClient implements HttpClient {
+
+    RestTemplate restTemplate;
+
+    @Autowired
+    public SpringHttpClient(RestTemplate restTemplate) {
+        this.restTemplate = restTemplate;
+    }
+
+    @Override
+    public <S extends Serializable, T extends Serializable> RestRequest<S, T> rest(Class<S> requestClass, Class<T> responseClass) {
+        return new SpringRestRequest<>(restTemplate, requestClass, responseClass);
+    }
+
+    @EqualsAndHashCode(callSuper = true)
+    @ToString
+    public static class SpringRestRequest<S extends Serializable, T extends Serializable> extends RestRequest<S, T> {
+
+        @ToString.Exclude
+        private RestTemplate restTemplate;
+
+        SpringRestRequest(RestTemplate restTemplate, Class<S> requestClass, Class<T> responseClass) {
+            super(requestClass, responseClass);
+            this.restTemplate = restTemplate;
+        }
+
+        @Override
+        public T sync(HttpMethod httpMethod) {
+            T responseObject;
+            org.springframework.http.HttpMethod springHttpMethod = org.springframework.http.HttpMethod.valueOf(httpMethod.toString());
```

Comment:
Is there a nicer way of writing these types?

Reply:
It's because I have a duplicate class name, so the 2nd one to be defined has to be fully declared to differentiate it.  The reason I defined a duplicate class is to keep the client decoupled from spring's, should we find out later the spring http client can't do something and we need to swap it for something else.  This is not completely unlikely as I've had cases before where a specific spring class was too limited and I had to use some other library for one small aspect.  It's also just basic OO to keep your APIs decoupled.